lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -1975,6 +1985,71 @@ void DifferentiableActivityInfo::propagateVaried(
   }
 }
 
+void DifferentiableActivityInfo::setUsefulAndPropagateToOperands(
+    SILValue value, unsigned dependentVariableIndex) {
+  // Skip already-useful values to prevent infinite recursion.
+  if (isUseful(value, dependentVariableIndex))
+    return;
+  if (value->getType().isAddress()) {
+    propagateUsefulThroughAddress(value, dependentVariableIndex);
+    return;
+  }
+  setUseful(value, dependentVariableIndex);
```

Comment:
Note: both `setUsefulAndPropagateToOperands` and `propagateUsefulThroughAddress` call `isUseful` and `setUseful`, which seems suboptimal. I tried removing `setUseful` from one of them but ran into infinite loops.

Related: I think `setUsefulAndPropagateToOperands` should be the primary entry point for propagating usefulness, so I eliminated most direct calls to `propagateUsefulThroughAddress`.